synapse/storage/persist_events.py

```diff
@@ -734,7 +754,86 @@ async def _get_new_state_after_events(
             state_res_store=StateResolutionStore(self.main_store),
         )
 
-        return res.state, None
+        # If the returned state matches the state group of one of the new
+        # forward extremities then we check if we are able to prune some state
+        # extremities.
+        if res.state_group and res.state_group in new_state_groups:
+            # We keep all the extremities that have the same state group, and
+            # see if we can drop the others.
+            new_new_extrems = {
+                e
+                for e in new_latest_event_ids
+                if event_id_to_state_group[e] == res.state_group
+            }
+
+            dropped = set(new_latest_event_ids) - new_new_extrems
+
+            logger.debug("Might drop events: %s", dropped)
+
+            # We only drop events if:
+            #   1. we're not currently persisting them;
+            #   2. they're not our own events (or are dummy events); and
+            #   3. they're either:
+            #       1. over N hours old and more than N events ago (we use depth
+            #          to calculate); or
+            #       2. we are persisting an event from the same domain.
+            #
+            # The idea is that we don't want to drop events that are "legitmate"
+            # extremities (that we would want to include as prev events), only
+            # "stuck" extremities that are e.g. due to a gap in the graph.
+            #
+            # Note that we either drop all of them or none of them. If we only
+            # drop some of the events we don't know if state res would come to
+            # the same conclusion.
+
+            drop = True
+            for ev, _ in events_context:
+                if ev.event_id in dropped:
+                    logger.debug("Not dropping as in to persist list")
+                    drop = False
+                    break
```

Comment:
Would it be clearer to return earlier in these spots? (Maybe not...)

Reply:
I don't think so